machinelearning/tensorflow.md

```diff
@@ -0,0 +1,349 @@
+# ML inference on Graviton CPUs with TensorFlow
+
+**Introduction**
+
+TensorFlow is an open-source software library for machine learning and artificial intelligence. It can be used across training and inference of deep neural networks. This document covers how to enable Tensorflow based machine learning inference on Graviton CPUs, what runtime configurations are important and how to debug any performance issues. The document also covers the instructions source builds to enable downstream features.
+
+# How to enable Tensorflow on Graviton CPUs
+To start with how to enable tensorflow inference on Graviton CPUs, there are multiple levels of software package abstractions available: Deep Lerning Amazon Machine Image (DLAMI), AWS Deep Learning Container (DLC), Docker hub container and Python wheel. Let's see how one can use each of them to get started.
+
+**Using AWS DLAMI**
+
+To launch a Graviton instance with a DLAMI via EC2 console:
+
+1. Click on "Community AMIs"
+2. Filter AMIs by clicking Ubuntu and Arm (64-bit) check boxes
+3. Search for "Deep Learning" in search field to list the current Ubuntu based DLAMIs"
+4. Select the AMI: "Deep Learning AMI Graviton Tensorflow <version>(Ubuntu 20.04) <yyyy/mm/dd>"
+
+Here is the awscli snippet to find the latest AMI:
+```
+aws ec2 describe-images --region us-west-2 --filters  Name=architecture,Values=arm64 Name=name,Values="*Deep Learning AMI*" Name=owner-alias,Values=amazon  --query 'Images[] | sort_by(@, &CreationDate)[-1] | ImageId'
+```
+
+Here is the awscli snippet to launch the same:
+```
+aws ec2 run-instances --region us-west-2 --image-id <ami-id from the above command>  --instance-type c7g.4xlarge --count 1  --key-name <key name> --subnet-id <subnet_id> --block-device-mappings 'DeviceName=/dev/sda1,Ebs={VolumeSize=256}'
+```
+
+Once Graviton instance is launched with the above DLAMI, the platform is ready for building Tensorflow based ML inference applications.The DLAMI comes with the additional libraries preinstalled, eg: rust compilers, transformers etc, to enable transformers based (eg: bert) inferencing applications.
+
+
+**Using Docker hub container**
+
+Launch a Graviton instance with an Ubuntu 20.04 AMI and instal docker using the below instructions.
+```
+#install and enable docker
+sudo apt update
+sudo apt install -y docker.io
+sudo systemctl enable docker && sudo systemctl start docker && sudo usermod -a -G docker ubuntu
+sudo chmod 666 /var/run/docker.sock
+
+# pull the tensorflow docker container with onednn-acl optimizations enabled
+docker pull armswdev/tensorflow-arm-neoverse
+
+# launch the docker image
+docker run -it --rm -v /home/ubuntu/:/hostfs armswdev/tensorflow-arm-neoverse
+```
+
+**Using Python wheel**
```

Comment:
it's super easy, do we want this one liner higher up?

Reply:
do you mean using the wheel? or neoverse container? I ordered them in the order of how complete the package is for someone to start ML app.